### ISSUE TITLE
Suggestions & comments

### ISSUE THREAD

**Issue by ozh** [ianbarber/Yourls-APC-Cache#1]
Ian,

Not really an "issue", but a few suggestions & comments
- as you may have seen in a tweet, a foolproof check would do no harm. Something like
  if( !function_exists( 'apc_exists' ) ) {
  yourls_die( 'You cannot run this plugin, noob!' );
  }
- Currently keyword infos are stored for a short time (APC_CACHE_TIMEOUT ie 120 seconds). I think these infos could be cached much more, as they're not supposed to change often.
- I think there might be another path to explore to further optimize things and reduce the number of queries. Currently on every redirection 4 SQL requests are issued (see http://code.google.com/p/yourls/issues/detail?id=279). You could easily cache the first one that fetchs various info from the DB, the main one being the list of active plugins.

**Comment by ianbarber** [comment]
Yep, I'll add a cache for the options, that makes sense and is straightforward. I can up the cache time, ideally I guess there would be a config admin thing, but that sounds like way too much effort :) I was thinking it might be nice to have the click log caching optional as well, but I can't think of a good way of doing the config other than suggesting people change a flag in the source.

**Comment by ianbarber** [comment]
OK, so I've added some config defines that can be put in config.php, and I've added a cache for the all_options, which I think should handle most of the options stuff!

**Comment by ianbarber** [comment]
Take a look at the new code if you get a sec - I think it should be a bit more complete, but there may be room for a new filter in get_all_options to give us a hook point to cache the result without having to duplicate the query code in the plugin.

**Comment by ozh** [comment]
I'm wondering something -- can't install APC on my dev box to confirm or unconfirm it but : yourls-loader.php fetches all options before including all plugins, therefore I'm not actually sure the shunt_get_all_options hook can be used. Can you confirm this?

**Comment by ianbarber** [comment]
That is a very good point - that code probably isn't doing much of anything. I should really verify these things before pushing them :) Hmm, could be tricky, though perhaps workaroundable by explicitly including a plugin reference in config.php. Not very nice though really.

**Comment by ozh** [comment]
Instead of including a file in the config.php (yuck) I can introduce a drop-in replacement file. Check file load-yourls.php: http://code.google.com/p/yourls/source/browse/trunk/includes/load-yourls.php#103
I think it makes sense that early layers such as DB engine or caching engine are loaded here.

In other words, there could be a if( file_exists( YOURLS_USERDIR.'/cache.php' ) ) {
    require_once( YOURLS_USERDIR.'/cache.php' );
}

Such a file would contain minimal code in your case: just load your plugin before anything goes further (and thus your plugin would be able to interrupt the get_all_options hook)

**Comment by ianbarber** [comment]
Yeah, that looks a lot cleaner a solution. If we could have that in, and something like     $ydb->option = yourls_apply_filter( 'get_all_options', $ydb->option ); in get_all_options that would make the implementation pretty decent.

**Comment by ozh** [comment]
See http://code.google.com/p/yourls/source/detail?r=644

**Comment by ianbarber** [comment]
OK, pushed an update. Checking it now, there are no DB accesses at all on most redirects, once the cache is filled. Working nicely!

**Comment by ozh** [comment]
Nitpicking: technically, the user directory and the plugin directory could be anywhere (user can define their location), so the include in the cache.php would be safer with:
include( YOURLS_PLUGINDIR . '/apc-cache/plugin.php' );
(which is not 100% foolproof yet since someone user change the plugin directory name -- let's assume they won't)

Very nice work overall !

**Comment by ianbarber** [comment]
Thanks Ozh! I've made that change. If you were interested, I've just pushed up a very simple plugin that just separates the views of shortened URLs for separate user accounts (e.g. so they don't show up in the admin or get access to the + stats page). 

https://github.com/ianbarber/Yourls-Separate-Users

It does require a couple of new filters - I sent a feature req with a patch (http://code.google.com/p/yourls/issues/detail?id=842) , though if you have a second to look and have suggestions on any different tactics would love to hear.
